mdata/cache/accnt/flat_accnt.go

```diff
@@ -0,0 +1,310 @@
+
+// it's easily possible for many events to happen in one request,
+// we never want this to fill up because otherwise events get dropped
+const eventQSize = 100000
+
+// Flat accounting
+//
+// Keeps track of the chunk cache size and in which order the contained
+// chunks have been used to last time. If it detects that the total cache
+// size is above the given limit, it feeds the least recently used
+// cache chunks into the evict queue, which will get consumed by the
+// evict loop.
+
+type FlatAccnt struct {
+	// total size of cache we're accounting for
+	total uint64
+
+	// metric accounting per metric key
+	metrics map[string]*FlatAccntMet
+
+	// the size limit, once this is reached we'll start evicting data
+	maxSize uint64
+
+	// a last-recently-used implementation that keeps track of all chunks
+	// and which hasn't been used for the longest time. the eviction
+	// function relies on this to know what to evict.
+	lru *LRU
+
+	// whenever a chunk gets evicted a job gets added to this queue. it is
+	// consumed by the chunk cache, which will evict whatever the jobs in
+	// evictQ tell it to
+	evictQ chan *EvictTarget
+
+	// a queue of add and hit events to be processed by the accounting.
+	// each add means data got added to the cache, each hit means data
+	// has been accessed and hence the LRU needs to be updated.
+	eventQ chan *FlatAccntEvent
+
+	stats       *Stats
+	statsTicker *time.Ticker
+	lastPrint   int64
+}
+
+type FlatAccntMet struct {
+	total  uint64
+	chunks map[uint32]uint64
+}
+
+// event types to be used in FlatAccntEvent
+const evnt_complt_met uint8 = 1
+const evnt_miss_met uint8 = 2
+const evnt_part_met uint8 = 3
+const evnt_hit_chnk uint8 = 4
+const evnt_add_chnk uint8 = 5
+
+type FlatAccntEvent struct {
+	t  uint8       // event type
+	pl interface{} // payload
+}
+
+type Stats struct {
+	// metric full hits, all requested chunks were cached
+	complt_met uint32
+
+	// metric complete misses, not a single chunk of the request was cached
+	miss_met uint32
+
+	// metric partial hits, some of the requested chunks were cached
+	part_met uint32
+
+	// new metrics added
+	add_met uint32
+
+	// metrics completely evicted
+	evict_met uint32
+
+	// chunk hits
+	hit_chnk uint32
+
+	// chunk adds
+	add_chnk uint32
+
+	// chunk evictions
+	evict_chnk uint32
+}
+
+// payload to be sent with an add event
+type AddPayload struct {
+	metric string
+	ts     uint32
+	size   uint64
+}
+
+// payload to be sent with a hit event
+type HitPayload struct {
+	metric string
+	ts     uint32
+}
+
+func NewFlatAccnt(maxSize uint64) Accnt {
+	accnt := FlatAccnt{
+		total:       0,
+		metrics:     make(map[string]*FlatAccntMet),
+		maxSize:     maxSize,
+		lru:         NewLRU(),
+		evictQ:      make(chan *EvictTarget, evictQSize),
+		eventQ:      make(chan *FlatAccntEvent, eventQSize),
+		stats:       &Stats{0, 0, 0, 0, 0, 0, 0, 0},
+		lastPrint:   time.Now().UnixNano(),
+		statsTicker: time.NewTicker(time.Second * 10),
+	}
```

Comment:
you don't need to specify attributes for which you want to initialize to the "zero value" (see https://golang.org/ref/spec#The_zero_value).

so total and stats can be left out. all those integers will be zeroes.